### ISSUE TITLE
Dagster uses an implausible number of postgresql connections

### ISSUE THREAD

**Issue by chrishiste** [dagster-io/dagster#20325]
### Dagster version

1.6.3

### What's the issue?

Hello,

We've encountered a significant issue with Dagster's interaction with PostgreSQL, leading to an exhaustion of the available connection slots. Despite increasing the connection limit from 100 to 1000, we are still facing the following error:

```
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) connection to server at "X.X.X.X", port 5432 failed: FATAL: remaining connection slots are reserved for non-replication superuser connections
```

This problem seems to arise when running Dagster jobs with multiple steps. However, the number of steps executed is not sufficient to justify the use of 1000 connections, suggesting an issue with how Dagster, specifically the `dagster/concurrency_key`, manages database connections. My hypothesis is that there might be a problem in how connections are created or reused, which becomes apparent under the load of multi-step jobs.

Can someone please look into this? Any guidance on diagnosing or resolving this issue would be greatly appreciated.

Thank you for your time.

### Deployment type

None

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by alangenfeld** [comment]
Are you able to query `pg_stat_activity` when this is happening? That may help point to a specific problematic interaction if there is a clear pattern. 

Investigating https://github.com/dagster-io/dagster/pull/20367 from skimming the code

**Comment by chrishiste** [comment]
@alangenfeld I will try next time it happens but it's not totally clear when it will be. I can see spikes in active connections for the last 30 days in my logs which is good because I will be able to confirm if your fix works.

**Comment by chrishiste** [comment]
It happened again in 1.7.1 so the fix wasn't effective. @alangenfeld what do you want from `pg_stat_activity`?

**Comment by alangenfeld** [comment]
> what do you want from pg_stat_activity?

`query` would probably be the most useful, but fine with `*`

**Comment by chrishiste** [comment]
@alangenfeld Sorry, I should have looked into what `pg_stat_activity` is before. This won't work in our scenario. This errors a bit unpredictable and depend on the load we have on our Dagster production instance, by the time I log into that instance the activity will likely be gone already. For historical analysis, I can setup Audit Logs would that be helpful?

**Comment by chrishiste** [comment]
@alangenfeld this happens quite often nowadays, I guess because we have more concurrency. I will be more than happy to provide help if you let me know what would be helpful to fix this issue. 1000 connections for 12 concurrent runs is definitely a bug, it's also adding strain on our MySQL instance because it has to manage so many connections. I had to double our vCPU count to make Dagit usable but still hitting 100% CPU usage consistently.

**Comment by chrishiste** [comment]
```python
import time

from dagster import (
    DynamicOut,
    DynamicOutput,
    graph_asset,
    op,
)


@op(out=DynamicOut())
def create_range(context):
    for i in range(100):
        yield DynamicOutput(mapping_key=str(i), value=str(i))


@op(
    tags={"dagster/concurrency_key": "serial"},
)
def sleep(context, value):
    time.sleep(2)
    return


@graph_asset(
    group_name="test",
)
def break_dagster():
    return create_range().map(sleep)

```

If you use this script and run like 5-10 of those concurrently with a Postgre that has the default 100 connections, you will see the issue. Each of those script open dozens of connections randomly.

**Comment by chrishiste** [comment]
Make sure to set the `serial` concurrency_key to 1 in the UI before hand

**Comment by efcaguab** [comment]
Facing the same issue. I tried a backfill with 6 concurrent runs. Each job's run has 7 steps. Steps were randomly failing with`psycopg2.OperationalError: connection to server at "XXX" (XXX), port 5432 failed: FATAL:  remaining connection slots are reserved for non-replication superuser connections` (full error at the end of this message).

Having a look at the db. The offending query is. 

```SQL
INSERT INTO
  event_logs (run_id,
    event,
    dagster_event_type,
    timestamp,
    step_key,
    asset_key,
    PARTITION)
VALUES
  ($1, $2, $3, $4::timestamp, $5, $6, $7) RETURNING event_logs.run_id,
  event_logs.id
```
The query is called almost 4000 times. Over the 15 mins that the backfill lasted. Which is pretty crazy. 

<img width="648" alt="image" src="https://github.com/dagster-io/dagster/assets/8033969/c40974bc-c4ec-4e7b-a22a-f62eb89aeb83">

I'm running dagster 1.7.4 with the `K8sRunLauncher`

```
psycopg2.OperationalError: connection to server at "XXX" (XXX), port 5432 failed: FATAL:  remaining connection slots are reserved for non-replication superuser connections


  File "/miniconda/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 145, in __init__
    self._dbapi_connection = engine.raw_connection()
                             ^^^^^^^^^^^^^^^^^^^^^^^
  File "/miniconda/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 3282, in raw_connection
    return self.pool.connect()
           ^^^^^^^^^^^^^^^^^^^
  File "/miniconda/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 449, in connect
    return _ConnectionFairy._checkout(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/miniconda/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 1263, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/miniconda/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 712, in checkout
    rec = pool._do_get()
          ^^^^^^^^^^^^^^
  File "/miniconda/lib/python3.12/site-packages/sqlalchemy/pool/impl.py", line 282, in _do_get
    return self._create_connection()
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/miniconda/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 390, in _create_connection
    return _ConnectionRecord(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/miniconda/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 674, in __init__
    self.__connect()
  File "/miniconda/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 900, in __connect
    with util.safe_reraise():
  File "/miniconda/lib/python3.12/site-packages/sqlalchemy/util/langhelpers.py", line 146, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/miniconda/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 896, in __connect
    self.dbapi_connection = connection = pool._invoke_creator(self)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/miniconda/lib/python3.12/site-packages/sqlalchemy/engine/create.py", line 643, in connect
    return dialect.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/miniconda/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 616, in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/miniconda/lib/python3.12/site-packages/psycopg2/__init__.py", line 122, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by efcaguab** [comment]
@alangenfeld Is the info above useful for tracing down the problem?

**Comment by alangenfeld** [comment]
Testing your script locally on master with postgres running in docker at 12 runs I am getting 10 - 15 concurrent connections.

I do see something unexpected in `pg_stat_activity` (our check for if the `concurrency_slots` table exists firing repeatedly) so I will send a fix out for that.

**Comment by chrishiste** [comment]
@alangenfeld exciting news, I will wait for that fix to land and will report back what happens

**Comment by emirkmo** [comment]
See old and still valid report on this issue: https://github.com/dagster-io/dagster/issues/8466

We have pgbouncer infront and that can handle the crazy amount of connections spammed, though the spam has not gone away. (This must be fixed in the cloud version to be honest? Maybe one can look there for inspiration.)

**Comment by caleboverman** [comment]
We're also experiencing this issue. A single materialization limited to 16 concurrent runs generates well over 2k postgres connections. We're trying to get pgbouncer or RDS proxies in place to help address the issue.

We began seeing this issue when migrating our k8s jobs to use k8s pipes for asset materialization. Prior to migrating we would simply launch jobs and not tie into any asset functionality and had no issues scaling the number of jobs we ran. Upon migrating to pipes and utilizing assets we quickly ran into the postgres connection issue.

**Comment by leo-schmidt** [comment]
Are there any updates on this issue?
We are also running into the same issue. We suddenly got 1000's of connections to the PostgresDB after deploying the DagsterDaemon. Our (temporary) solution was to shut down the Daemon deployment and use GraphQL to start jobs, however this means we can not use any sensors or schedules.

**Comment by smackesey** [comment]
@alangenfeld do you know if anything relevant has changed here?

**Comment by chrishiste** [comment]
This is still an issue

Related threads: 
https://github.com/dagster-io/dagster/issues/8466
https://github.com/dagster-io/dagster/issues/16138

**Comment by tobuhl** [comment]
We are experiencing the same behaviour. Is there any news?

**Comment by YasmineMo** [comment]
I am facing the same issue even with a pgbouncer. I am using Dagster OSS 1.7.8 and K8s launder. Without the use of pgbouncer, I could barely run 9 dagster jobs.. it went well for sometime (at least I thought so), now it is crazy again with jobs takes hours to finish.

**Comment by mcguipat** [comment]
I'm encountering the same issues using dagster/dagster-webserver 1.9.13 and dagster-postgres 0.25.13

The issue is manifesting itself with the following exception: `DagsterPosgresException("too many retries for DB connection")`

We've tried using PgBouncer to pool connections but this does not resolve the issue.  

Is there any idea around the possible cause or any fix being explored?  I cannot understand how this is not creating a significant issues for many other users.

**Comment by abhinavDhulipala** [comment]
Are there any immediate workarounds? Like going through the webserver pooler or having an intermediate pg_bouncer pod. We already have an RDS proxy and are still hitting this with the kubernetes_job_executor.

**Comment by mcguipat** [comment]
> Are there any immediate workarounds? Like going through the webserver pooler or having an intermediate pg_bouncer pod. We already have an RDS proxy and are still hitting this with the kubernetes_job_executor.

One thing I’ve found to eliminate the previous issues I was having with this was to set `use_threads` to false in `QueuedRunCoordinator`

**Comment by abhinavDhulipala** [comment]
This would seem to be helpful in reducing the amount of runs that get dispatched at once. I don't think that's our problem. We need the concurrency because we need the throughput of work. Our problem is induced by just scaling up the max parallel runs, I don't think the daemon dispatch is our problem.

**Comment by abhinavDhulipala** [comment]
We were running into DNS errors. Here is our exact error logs:

```bash
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not translate host name "dagster-db-proxy.XXXX.us-west-2.rds.amazonaws.com" to address: Temporary failure in name resolution
(Background on this error at: https://sqlalche.me/e/20/e3q8)
```


 This was because VPC's have a DNS [limit](https://docs.aws.amazon.com/whitepapers/latest/hybrid-cloud-dns-options-for-vpc/constraints.html) 1024 packets per sec and we use the kubernetes job executor with a lot of parallel jobs. We got around this by deploying node local [DNS](https://kubernetes.io/docs/tasks/administer-cluster/nodelocaldns/) and then modifying our managed node groups with the following:

```yml

apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig
metadata:
  name: cluster-name
  region: our-region-1
vpc:
# ... vpc detais
managedNodeGroups:
  - name: node-group-name
    amiFamily: AmazonLinux2
    ami: ami-id
    # had to add this to our nodegroup configs
    overrideBootstrapCommand: |
      #!/bin/bash
      set -ex
      /etc/eks/bootstrap.sh cluster-name --dns-cluster-ip 169.254.20.10. # this is the default, substitute if using non-default node dns ip.
```

We published with `eksctl create nodegroup --config-file config-file.yml` and that solved our DNS problems at scale. 

We then slowly cordon'd and transitioned to nodegroups with dns cluster ip pointing to our nodelocal dns cache. This was a problem because CoreDNS still reaches out to route53 for "external" ip's. So everything that isn't a k8s service within your eks cluster. 

Another solution would also be to use a k8s local pgbouncer service within your k8s  cluster (instead of or along side RDS proxy). Although I haven't tried this, I think this could work as well if bootstrapping your nodegroups is out of the question. 

To summarize, our setup for scale is NodeLocalDNS + Calico Networking (to avoid ipv4 exhaustion) + RDS (Behind an RDS proxy). 

Having said all this, application side pooling should happen. We shouldn't really even be in position where we are issueing this many inflight requests. The design of the k8s job executor should interact with the launcher pod and pool it's connections there or come with another service that can serve all runs and serve as an api/pooler.

Hope this helps someone else!

**Comment by emirkmo** [comment]
> I'm encountering the same issues using dagster/dagster-webserver 1.9.13 and dagster-postgres 0.25.13
> 
> The issue is manifesting itself with the following exception: `DagsterPosgresException("too many retries for DB connection")`
> 
> We've tried using PgBouncer to pool connections but this does not resolve the issue.
> 
> For a concurrent run of 25 jobs where the job has 5 assets and no logging within the assets, we see ~87k queries being executed which in itself looks like the problem. I would think there would be some sort of batching of record insertion but it doesn't seem to be the case.
> 
> Is there any idea around the possible cause or any fix being explored? I cannot understand how this is not creating a significant issues for many other users.

If you configure PgBouncer correctly, it should mitigate the issue for you.

**Comment by janosroden** [comment]
> > I'm encountering the same issues using dagster/dagster-webserver 1.9.13 and dagster-postgres 0.25.13
> > The issue is manifesting itself with the following exception: `DagsterPosgresException("too many retries for DB connection")`
> > We've tried using PgBouncer to pool connections but this does not resolve the issue.
> > For a concurrent run of 25 jobs where the job has 5 assets and no logging within the assets, we see ~87k queries being executed which in itself looks like the problem. I would think there would be some sort of batching of record insertion but it doesn't seem to be the case.
> > Is there any idea around the possible cause or any fix being explored? I cannot understand how this is not creating a significant issues for many other users.
> 
> If you configure PgBouncer correctly, it should mitigate the issue for you.

I tried that a year ago, but dagster webserver sets session variables, which is allowed only in session pool mode. Therefore, I configured pgbouncer to use session pooling, which was not enough for me (I had 5 free slots or so for the db server). I didn't investigate further, but I reached my connection limit with a few parallel runs.

**Comment by emirkmo** [comment]
Don't get me wrong, this is a legitimate bug that Dagster haven't addressed in OS version for a few years.

However, we've been using pgbouncer to work around it for a few years as well. Connection pooling with a high number of connections set seems to work.

**Comment by joshuataylor** [comment]
I know that Dagster now has support for RDS Proxy, has anyone who has been using RDS Proxy for this issue seen a benefit?

I really don't want to have to deploy PGBouncer just for this issue, it's fine if I have to, just a bit of a PITA, as our AWS setup is pretty simple, using ECS with Dagster.

**Comment by laegsgaardTroels** [comment]
Using postgres with pgbouncer and `k8s_job_executor`. 

Postgres is handling it fine. The k8s CoreDNS is often throttled because of the many requests Dagster makes to postgres.

Is it on the roadmap to reduce the many requests made to storage?

And do you know if the problem is isolated to the `k8s_job_executor`?

**Comment by sqkkyzx** [comment]
I tried setting PG_DB_CONN_STRING to three different users for dagster_daemon, dagster_webserver and dagster_code to see what was making the excessive database connections, and found that it was dagster_daemon.

Using:

```sql
SELECT   
    u.usename,  
    s.query,   
    s.calls,   
    s.total_exec_time,   
    s.mean_exec_time   
FROM pg_stat_statements s  
JOIN pg_user u ON s.userid = u.usesysid  
ORDER BY s.calls DESC  
LIMIT 5;
```

I discovered that the following statement:

```sql
SELECT pg_catalog.pg_class.relname   
FROM pg_catalog.pg_class JOIN pg_catalog.pg_namespace ON pg_catalog.pg_namespace.oid = pg_catalog.pg_class.relnamespace   
WHERE pg_catalog.pg_class.relkind = ANY (ARRAY[$1, $2]) AND pg_catalog.pg_class.relpersistence != $3 AND pg_catalog.pg_table_is_visible(pg_catalog.pg_class.oid) AND pg_catalog.pg_namespace.nspname != $4
```

has a total_exec_time of 35411 and was executed 69,949 times — absolutely excessive.

In another misconfiguration I found that SQLAlchemy was using poolclass=NullPool.

So PgBouncer may indeed be necessary. I previously introduced PgBouncer and things quieted down for a while; when PgBouncer was removed the excessive activity started again.

**Comment by VinzenzKlass** [comment]
@sqkkyzx We observed almost identical behaviour with these reflection requests hitting our postgres almost constantly. Could you perhaps check if #33413 solves your issues as well? I think it's quite likely that there are more occurances of heavily repeated reflection in the code, so I may need to still ammend my PR.

@alangenfeld Since you have looked at this and created this PR #22095, could you take a look at this and see if this makes sense to you? Happy to put some more time into the PR draft to get this resolved ASAP.

**Comment by 0xfabioo** [comment]
Is there any update on this issue? We’re using burstable DB instances in our setup, and this behavior is causing significantly higher credit consumption than expected.

Any guidance or workaround would be greatly appreciated.

**Comment by TabLand** [comment]
Hi Team, 

Can we get some movement on this issue please? Easily one of the most obvious ways dagster can be made more efficient. 

We could never justify writing to a file by creating fresh file handles and then only writing a single character per handle? Batched transactions and flush / commit operations exist for a reason. We can make this behaviour configurable if anyone has a genuine concern for full idempotence. 

I recently compared some run enqueue times for dagster on postgres versus SQLite and we're seeing 30 seconds for the former compared to roughly 3-6 seconds for the latter, with first log event happening near 1 minute for postgres compared to roughly 11 seconds for SQLite. We can also see spikes of 3.5k connections at busy times, and 33k connections per ~day~ hour and around 700k per day . We only have 50 jobs with under 1000 runs per day. On the database server side, each connection object spawns a worker process or thread that inevitably has a footprint.

This isn't an option for batch workflows where minutes and seconds matter. I hope this isn't a general trend amongst OSS towards wasting compute, memory and network as a way of appeasing cloud companies...?

I can see some PR's have been shared recently, I will endeavour to test them and feedback within a fortnight and then escalate this issue via adjacent channels.

**Comment by TabLand** [comment]
Further update:

We're seeing total of 491098 IO operations per day, based on an average of 5.703 IOPS. That's technically < 1 IOP per connection as it probably excludes memory cache hits. Total reads = 463 MB; Total writes = 8.07 GB. Per connection, that results in around 694 bytes read, and 12.25 kb written. Would like to see numbers from other users.

**Comment by TabLand** [comment]
I've had some decent results with the following changes in dagster 1.7.7:

```python
#dagster_postgres\utils.py:156
conn: psycopg2.extensions.connection = None
@contextmanager
def create_pg_connection(
    engine: sqlalchemy.engine.Engine,
) -> Iterator[Connection]:
    check.inst_param(engine, "engine", sqlalchemy.engine.Engine)
    global conn
    try:
        if conn != None and not conn.closed and conn.status == psycopg2.extensions.STATUS_READY:
            yield conn
        else:
            # Retry connection to gracefully handle transient connection issues
            conn = retry_pg_connection_fn(engine.connect)
            yield conn
    except Exception as e:
        print("Error connecting to Postgres")
        raise DagsterPostgresException("Failed to connect to Postgres") from e
    finally:
        if conn != None and not conn.closed:
            conn.close()

#dagster\_core\events\__init__.py:247
# These are the only events currently supported in `EventLogStorage.store_event_batch`
BATCH_WRITABLE_EVENTS = {
    DagsterEventType.ASSET_MATERIALIZATION,
    DagsterEventType.ASSET_OBSERVATION,
    DagsterEventType.LOGS_CAPTURED, #For faster UI log writing
}
```

I also opted to set `DAGSTER_EVENT_BATCH_SIZE=50` in my `.env` file. You can find the documentation for that here, worth trying this as a tiny stop gap in case this helps alleviate some performance bottlenecks for wider users: 

>[experimental] Single-run backfills now support batched inserts of asset materialization events. This is a major performance improvement for large single-run backfills that have database writes as a bottleneck. The feature is off by default and can be enabled by setting the DAGSTER_EVENT_BATCH_SIZE environment variable in a code server to an integer (25 recommended, 50 max). It is only currently supported in Dagster Cloud and OSS deployments with a postgres backend.

https://docs.dagster.io/about/changelog#:~:text=%5Bexperimental%5D%20Single%2Drun,a%20postgres%20backend.

Btw, I can't see any checks for this max batch size of 50. Would it be good to have a slightly higher option available just for front end logging...?

Will look to create a fresh PR for the latest version after testing this one for my needs.

**Comment by TabLand** [comment]
Please take the above initial analysis with a grain of salt. I have since learnt that the `conn` object is provided by sqlalchemy, not psycopg2 which has connection pooling on the client side:

https://docs.sqlalchemy.org/en/21/core/pooling.html

I now feel the issue might be with these lines here, which forces sqlalchemy to create lots of new connections unnecessarily:

```python
#dagster_postgres\utils.py
    finally:
            conn.close()
```

Perhaps we just need to commit rather than close the connection object in order to support the index transactions, which are seldom used. I'm not convinced these are being used for records being inserted into the event log; I couldn't find the handoff between the configured python logger and the `SqlEventLogStorage` class.

Lastly - the following could be a smoking gun too:

https://github.com/dagster-io/dagster/blob/94b62a14e5f715b2d8923b5de5aafc1b493b441f/python_modules/libraries/dagster-postgres/dagster_postgres/event_log/event_log.py#L90-L93

**Comment by joshuataylor** [comment]
The last time we tried to do any kind of tracing with DataDog and Dagster, we had multiprocessing issues, I'm not sure if it's been resolved yet (it looks like it hasn't), but it'll be interesting to see if otel-related could help, even in development.

Goodluck - debugging/tracing/profiling Dagster is tricky, especially with multiprocess threads, ensure you do things like:

```
    my_job = define_asset_job(
        executor_def=dagster_execution_mode(),
    )
```

**Comment by jmoldow** [comment]
> I discovered that the following statement:
> 
> SELECT pg_catalog.pg_class.relname   
> FROM pg_catalog.pg_class JOIN pg_catalog.pg_namespace ON pg_catalog.pg_namespace.oid = pg_catalog.pg_class.relnamespace   
> WHERE pg_catalog.pg_class.relkind = ANY (ARRAY[$1, $2]) AND pg_catalog.pg_class.relpersistence != $3 AND pg_catalog.pg_table_is_visible(pg_catalog.pg_class.oid) AND pg_catalog.pg_namespace.nspname != $4
> 
> has a total_exec_time of 35411 and was executed 69,949 times — absolutely excessive.

I've also observed that excess.  #33820 would solve that problem.

I don't know if it would solve the entirety of the problem of high number of connections, but I bet it would help significantly.
